rfcs/0011-ban-scoring.md

```diff
@@ -0,0 +1,159 @@
+the absence of observable misbehavior.
+
+In the Bitcoin C++ implementation, there's an API "Misbehaving" to
+increment the ban score, and set a ban flag if the score exceeds the
+threshold. In some cases, the ban score is incremented by a fixed
+amount. In other cases, the score is incremented by the value of a
+"DoS" value associated with a peer. There's a separate API for
+incrementing the DoS value when rejecting a transaction.
+
+At the time of writing (commit d6e700e), the Misbehaving API is called for certain 
+behaviors with fixed ban score increments. These behaviors have a score of 100,
+resulting in an immediate ban:
+
+ - invalid block, invalid compact block, nonmatching block transactions
+ - invalid Bloom filter version, too-large Bloom filter, 
+ - too-large data item to add to Bloom filter, add to missing Bloom filter
+ - invalid orphan tx, out-of-bounds tx indices
+ 
+One misbehavior has a score of 50: 
+
+ - message too large for buffer
+
+For a score of 20:
+
+ - too many unconnecting headers
+ - too-big message "addr", "inventory", or "getdata" message sizes
+ - too many headers, non-continuous headers
+
+And a score of of 1:
+
+ - missing "version" or "verack" messages, duplicate "version" message
+
+In several places in the code, the "DoS" value is checked, and if
+positive, it's added to the ban score.  The DoS score can be
+incremented when a transaction is rejected. Many reasons for rejecting
+a transaction add 100 to the DoS value (too many to enumerate
+here). Those include items like transaction fees out of range, missing
+or already-spent inputs, incorrect proof of work, and an invalid
+Merkle root.
+
+Some reasons for rejecting a transaction increment the DoS value a
+lesser amount. For instance, and invalid hash such that the proof of
+work fails, increments DoS by 50. A previous-block-not-found rejection
+increments DoS by 10. In several cases, transactions are rejected, but
+the DoS is not incremented, such as the "mempool-full" condition.
+
+# Coda at the moment
+
+Bitcoin is a mature codebase, so there are many places where ban scoring has been 
+used. Nonetheless, Bitcoin uses a relatively coarse ban scoring system; only 
+a few ban score increment values are used. In Coda, we could reify scores into a 
+datatype:
+
+```ocaml
+  type ban_score = 
+    | Ban_score_severe
+	| Ban_score_moderate
+	| Ban_score_trivial
+```
+A slightly finer gradation could be used, if desired. The blacklisting system 
+could translate these constructors into numerical scores. Let's call these
+constructors SEV, MOD, and TRV.
+
+In a number of places, the Coda codebase has comments indicating that 
+a peer should be punished, either via a `TODO` or call to `Logger.faulty_peer`.
+Let's classify those places where punishment has been flagged, and annotate 
+them with suggested constructors:
+
+  - in `bootstrap_controller.ml`, for bad proofs (SEV), and a validation error when 
```

Comment:
It'd be best if we separated the set of offenses in two.

1. Absent hardware failures, an honest peer never does this. Should be an insta-ban, since it's a clear signal the peer is faulty or malicious.

2. Honest peers may do this if they are out of sync with us. Set punishment level based on the cost to us of receiving/processing the bad message. Not sure how that converts into a number

Reply:
Maybe there's still room for trivial infractions, the kind where Bitcoin increments the ban score by one, like a duplicate version message. We could see that kind of thing if there are multiple Coda implementations.